plugins/colorbutton/plugin.js

```diff
@@ -370,6 +414,152 @@ CKEDITOR.plugins.add( 'colorbutton', {
 			}
 		}
 
+		function renderContentColors( options ) {
+			// This function is called on each dialog opening.
+			var contentColorsRow = options.contentColorsRow,
+				contentColorsLabel = options.contentColorsLabel,
+				cssAttribute = options.cssAttribute,
+				clickFn = options.clickFn,
+				type = options.type,
+				colorSpans = editor.editable().find( 'span[style*=' + cssAttribute + ']' ).toArray(),
+				colors = [],
+				colorOccurrences = {},
+				sortedColors = [],
+				colorsPerRow = config.colorButton_colorsPerRow || 6,
+				tilePosition = 1;
+
+			// It's more reliable and easier to delete exisiting color tiles and recreate them than to manipulate the old ones.
+			CKEDITOR.tools.array.forEach( contentColorsRow.find( 'td' ).toArray(), function( node ) {
+				node.remove();
+			}, 0, 1 );
+
+			if ( !colorSpans.length ) {
+				contentColorsLabel.hide();
+				return;
+			}
+
+			CKEDITOR.tools.array.forEach( colorSpans, function( span ) {
+				// colorSpans may contain spans with background color when they are not necessary, so here they are being filtered out.
+				var spanColor = span.getStyle( cssAttribute );
+
+				if ( spanColor ) {
+					colors.push( normalizeColor( spanColor ).toUpperCase() );
+				}
+			} );
+
+			if ( !colors.length ) {
+				contentColorsLabel.hide();
+				return;
+			}
+
+			CKEDITOR.tools.array.forEach( colors, function( color ) {
+				if ( color in colorOccurrences ) {
+					colorOccurrences[ color ]++;
+				} else {
+					colorOccurrences[ color ] = 1;
+				}
+			} );
+
+			for ( var color in colorOccurrences ) {
+				sortedColors.push( [ color, colorOccurrences[ color ] ] );
+			}
+			sortedColors.sort( function( a, b ) {
+				return b[ 1 ] - a[ 1 ];
+			} );
+
+			// Trim the color array to the row size.
+			if ( sortedColors.length > colorsPerRow ) {
+				sortedColors.splice( colorsPerRow - 1, sortedColors.length - colorsPerRow );
+			}
+
+			CKEDITOR.tools.array.forEach( sortedColors, function( color ) {
+				color[ 2 ] = editor.lang.colorbutton.colors[ color[ 0 ] ] || color[ 0 ];
+			} );
+
+			CKEDITOR.tools.array.forEach( sortedColors, function( color ) {
+				// Unfortunately CKEDITOR.dom.element.createFromHtml() doesn't work for table elements,
+				// so table cell has to be created separately.
+				var colorTile = new CKEDITOR.dom.element( 'td' );
+
+				colorTile.setHtml( generateTileHtml( {
+					colorLabel: color[ 2 ],
+					clickFn: clickFn,
+					colorCode: color[ 0 ],
+					type: type,
+					position: tilePosition++,
+					setSize: sortedColors.length
+				} ) );
+
+				contentColorsRow.append( colorTile );
+			} );
+
+			contentColorsLabel.show();
+		}
+
+		function addCustomColorToPanel( options ) {
```

Comment:
I won't go with #separateFunctionX here, but you already know what we could improve here.